### PR TITLE
Add platform support doc for 3ds

### PR DESCRIPTION
**Pull request by ian-h-chamberlain**
Ref https://github.com/Meziu/rust-horizon/issues/6

@AzureMarker @Meziu

I tried to avoid too much mention of Nintendo specifically and limited mentions of devkitARM as best I could, but it seems a bit hard to avoid completely since it's needed to actually build anything.

Let me know if you think anything is too wordy, or you think something should be changed or omitted. I think it's important to get this right since it's documentation and also one of the most likely entry points for users (or new contributors) of `ctru-rs`!

[The template for these docs](https://doc.rust-lang.org/nightly/rustc/platform-support/TEMPLATE.html) for reference.

### CODE REVIEW

File: src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+## Target maintainers
+
+- [@Meziu](https://github.com/Meziu)
+- [@AzureMarker](https://github.com/AzureMarker)
+- [@ian-h-chamberlain](https://github.com/ian-h-chamberlain)
+
+## Requirements
+
+This target is cross-compiled. `#![no_std]` crates should work without much
+additional effort.
```
Comment (ian-h-chamberlain) [inline src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md:17]: Oops, I just realized this is probably not completely true, do `no_std` crates still need to use `-Zbuild-std=core` or something like that?

Comment (AzureMarker) [reply src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md:17]: Yeah I think they need core to be built via build-std.

File: src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+fn main() {
+    ctru::init();
+}
+```
+
+## Testing
+
+Binaries built for this target can be run in an emulator (most commonly
+[Citra](https://citra-emu.org/)), or sent to a device through
+the use of a tool like devkitARM's `3dslink`.
```
Comment (AzureMarker) [inline src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md:93]: Maybe also note that you can copy them to the SD card, which is what I was doing initially :).

File: src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+The `cargo-3ds` tool mentioned in "Building Rust programs" supports the use of
+`3dslink` with `cargo 3ds run`. The default Rust test runner is not supported, but
```
Comment (AzureMarker) [inline src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md:96]: The default test runner is only not supported because it may potentially fork right? Maybe we could patch that in one of our rustc PRs to make it work out of the box.

File: src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+[custom test frameworks](https://doc.rust-lang.org/beta/unstable-book/language-features/custom-test-frameworks.html)
+can be used with `cargo 3ds test` to run unit tests on a device.
+
+The Rust `library/std` test suite is not yet supported.
+
+## Cross-compilation toolchains and C code
+
+C code can be built for this target using the aforementioned devkitARM toolchain.
```
Comment (AzureMarker) [inline src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md:104]: Maybe add or move some of the devkitpro description here, since this section seems to be the right place for that information.

**Review by AzureMarker — COMMENTED** [review]
Thanks for working on this!

**Comment by AzureMarker** [comment]
Oh, and also maybe we should explicitly mention that this is for homebrew?

File: src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md (hunk 1 above)
Comment (Meziu) [reply src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md:17]: Plus, the C RunTime is available only in devkitARM, so that’s another thing.

(Now that I think about it, mention that the platform doesn’t support dynamic linking).

File: src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+- [@ian-h-chamberlain](https://github.com/ian-h-chamberlain)
+
+## Requirements
+
+This target is cross-compiled. `#![no_std]` crates should work without much
+additional effort.
+
+`std` is partially supported, but mostly works. Some APIs are unimplemented
+and will simply return an error, such as `std::process`. An allocator is provided
+by default.
```
[the remark this answers is not in the document]
Comment (Meziu) [reply src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md:21]: The default one in the `std` for unix just works, I believe the guys before us had to reimplement it because they didn’t have access to `posix_memalign`.

File: src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+In order to support `std` APIs, binaries must be linked against the open-source
+[devkitARM toolchain](https://devkitpro.org/wiki/Getting_Started), notably
+using its `arm-none-eabi-gcc` as the linker.
+
+Additionally, some helper crates provide implementations of some `libc` functions
+use by `std`. These, or an alternate implementation of the relevant functions,
+are required to use `std`:
+
+- [`pthread-3ds`](https://github.com/Meziu/pthread-3ds) provides pthread APIs for `std::thread`.
+- [`linker-fix-3ds`](https://github.com/Meziu/rust-linker-fix-3ds) fulfills some other missing libc APIs.
```
[the remark this answers is not in the document]
Comment (Meziu) [reply src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md:32]: Maybe `ctru-sys`, if anything, though this falls in the considerations I had before…

**Comment by Meziu** [comment]
> Oh, and also maybe we should explicitly mention that this is for homebrew?

We absolutely should, stating we also aren’t affiliated with Nintendo.

File: src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md [hunk 7]
```diff
@@ -0,0 +1,104 @@
+# `armv6k-nintendo-3ds`
+
+**Tier: 3**
+
+The Nintendo 3DS platform, which has an ARMv6K processor, and its associated
+operating system (`horizon`).
```
[the remark this answers is not in the document]
Comment (ian-h-chamberlain) [reply src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md:6]: True, I was sort of following the format of some other docs, e.g. https://github.com/rust-lang/rust/blob/master/src/doc/rustc/src/platform-support/wasm64-unknown-unknown.md?plain=1#L5

I could try to reword, maybe just "This target is ..." ?

File: src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (ian-h-chamberlain) [reply src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md:32]: Since `ctru-sys` is a dependency of these crates, I think it's okay to omit here, since it's not required as a direct dependency, but I can add it if it seems relevant.

File: src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+Binaries built for this target should be compatible with most variants of the
+3DS (and 2DS) hardware and firmware, but testing is limited and some versions may
+not work correctly.
+
+This target generates binaries in the ELF format.
+
+## Building the target
+
+You can build Rust with support for the target by adding it to the `target`
```
[the remark this answers is not in the document]
Comment (ian-h-chamberlain) [reply src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md:42]: Actually I should test this too... I have built `std` for the target like this but never did an end-to-end since `cargo 3ds` is pretty convenient and the `std` code was changing pretty often.

File: src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md (hunk 3 above)
Comment (ian-h-chamberlain) [reply src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md:96]: Yeah, there are a handful of options that are hardcoded in our custom test runner right now, but maybe it would be possible to `cfg` them properly for the given target to make it work. As far as I can tell there's not much precedent for something like that, but maybe we could still do it. A future investigation to have!

File: src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (AzureMarker) [reply src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md:39]: True, I guess it should technically compile std with just those two (and dependencies).

File: src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (AzureMarker) [reply src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md:6]: It looks OK now, I'll resolve this.

File: src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md [hunk 9]
```diff
@@ -90,15 +98,24 @@ fn main() {
...
 
 ## Cross-compilation toolchains and C code
 
-C code can be built for this target using the aforementioned devkitARM toolchain.
+C code can be built for this target using the
+[devkitARM toolchain](https://devkitpro.org/wiki/Getting_Started).
+This toolchain provides `arm-none-eabi-gcc` as the linker used to link Rust
+programs as well.
+
+The toolchain also provides a `libc` implementation, which is required by `std`
```
Comment (AzureMarker) [inline src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md:119]: Maybe note that this libc implementation doesn't implement everything rustc expects?

File: src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (AzureMarker) [reply src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md:49]: I ran into some small issues with this. At first just some warnings-turned-errors, which I fixed in https://github.com/AzureMarker/rust-horizon/commit/96295e829409c6ac4a8d92b2bbfd5bbfe0464694 and https://github.com/AzureMarker/rust-horizon/commit/b51f37d1611a156d1333593112bf15fb4b080d5c. But then actually compiling something would give me the following link error (simplified paths a bit):
```
/opt/devkitpro/devkitARM/arm-none-eabi/bin/ld: error: hello_world-9db584d0ec6e1906.elf uses VFP register arguments, libcompiler_builtins-e5be153354215cbd.rlib(aeabi_div0.o) does not
          /opt/devkitpro/devkitARM/arm-none-eabi/bin/ld: failed to merge target specific data of file libcompiler_builtins-e5be153354215cbd.rlib(aeabi_div0.o)
```

File: src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md (hunk 9 above)
Comment (ian-h-chamberlain) [reply src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md:119]: Sure, I can make sure I mention that that for the next commit I push.

File: src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (ian-h-chamberlain) [reply src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md:49]: I forgot about the warnings-as-errors (we were going to have to fix those to merge upstream anyway, so thanks for fixing!).

As for the linker errors, I saw the same while trying to get `./x.py test` to build for the target, which makes me think if we could figure out how to solve it here we would be closer on that front as well. I asked on Zulip if anyone knew how to help, but unfortunately got no responses... 

I had also seen this kind of link error when hacking around previously (maybe when working on https://github.com/Meziu/libc/pull/8 ?), but I cannot quite remember how I got past it back then. I did find this issue, but it's a bit old and doesn't seem 100% relevant for our scenario, since `build-std` works correctly: https://github.com/rust-embedded/book/issues/255

Also found [this stackoverflow](https://stackoverflow.com/questions/9753749/arm-compilation-error-vfp-registers-used-by-executable-not-object-file), which is jogging my memory that perhaps I exported `CFLAGS` or `CXXFLAGS` (or both) somewhere to get around it (forcing to build with the same flags as in https://github.com/AzureMarker/rust-horizon/blob/horizon-std/compiler/rustc_target/src/spec/armv6k_nintendo_3ds.rs . 

`libc-test` uses the `cc-rs` crate, though, which may not be applicable for `compiler_builtins`, so maybe something else is at play, but that might be worth a try? Or perhaps there's some `config.toml` values that can be tweaked to get the behavior needed here?

[the remark this answers is not in the document]
Comment (Meziu) [reply src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md:49]: Yeah the `VFP register arguments` is something I had to fix too. Putting those vars in the target specification fixed it. I believe it has to do with the size/instructions available to the CPU, which the other linking processes (like `libctru` or whatever happens in `arm-none-eabi-gcc`) don’t find compatible in some specific cases. I even tried to get down to only one of those flags being the problem, though I don’t remember which, and I judged it’d be better to keep all of them anyways, for future problems may arise.

**Comment by ian-h-chamberlain** [comment]
Btw @FenrirWolf if you are still having trouble getting the environment going, maybe this will help? If not, it probably means we have a documentation gap and should address that anyway.

File: src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (AzureMarker) [reply src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md:49]: > ```shell
> $ export CFLAGS_armv6k_nintendo_3ds='-mfloat-abi=hard -mtune=mpcore -mtp=soft -march=armv6k'
> ```

Setting this before `./x.py build` seems to work. I didn't even set it before `cargo build --target armv6k-nintendo-3ds --example hello-world` in ctru-rs and it worked (separate terminal). I was able to execute it just fine on my 3DS, after converting to 3dsx.

[the remark this answers is not in the document]
Comment (ian-h-chamberlain) [reply src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md:49]: Nice! I will add to the docs about `Building Rust for the Target`. I think after digging a bit more, this is a gap in the default flags set by `cc-rs`, since it doesn't have knowledge about the target architecture. We could probably deliver a fix upstream there to resolve it, but exporting `CFLAGS_armv6k_nintendo_3ds` seems like the best workaround meanwhile.

**Review by AzureMarker — APPROVED** [review]
